src/sentry/integrations/repositories.py

```diff
@@ -2,21 +2,61 @@
 
 from sentry.constants import ObjectStatus
 from sentry.models import Repository
+from sentry.shared_integrations.exceptions import ApiError
 
 
 class RepositoryMixin(object):
     # whether or not integration has the ability to search through Repositories
     # dynamically given a search query
     repo_search = False
 
-    def get_stacktrace_link(self, repo, path, version):
+    def format_source_url(self, repo, filepath, branch):
+        """
+        Formats the source code url used for stack trace linking.
+        """
+        raise NotImplementedError
+
+    def check_file(self, repo, filepath, branch):
+        """
+        Calls the client's `check_file` method to see if the file exists.
+        Returns `True` if it does, and `False` if we 404.
+
+        So far only GitHub and GitLab have this implemented, both of which
+        give use back 404s. If for some reason an integration gives back
+        a different status code, this method could be overwritten.
+
+        repo: Repository (object)
+        filepath: file from the stacktrace (string)
+        branch: commitsha or default_branch (string)
+        """
+        try:
+            self.get_client().check_file(repo, filepath, branch)
+        except ApiError as e:
+            if e.code != 404:
+                raise
+            return False
+
+        return self.format_source_url(repo, filepath, branch)
+
+    def get_stacktrace_link(self, repo, filepath, default, version):
         """
         Handle formatting and returning back the stack trace link if the client
         request was successful.
 
+        Uses the version first, and re-tries with the default branch if we 404
+        trying to use the version (commit sha)
+
         If no file was found return `None`, and re-raise for non "Not Found" errors
+
         """
-        raise NotImplementedError
+        if version:
+            source_url = self.check_file(repo, filepath, version)
+            if source_url:
+                return source_url
+
+        source_url = self.check_file(repo, filepath, default)
+
+        return source_url if source_url else None
```

Comment:
@MeredithAnya we can probably just return `source_url` here with the change that I suggested above